feat: Merge Multiple Polygons in a COCO Dataset Annotation

DancinParrot opened the pull request:
# Description

This PR attempts to address issue #1209. Within an appropriately structured COCO dataset, an annotation may contain a segmentation list with multiple nested lists of vertices. This results in supervision terminating prematurely and raising a ValueError.

The solution as discussed in the issue includes the creation of binary masks for all polygons within the segmentation list and the usage of numpy's logical_or() function to merge all masks which is used in the `coco_annotations_to_masks` function.

## Type of change

Please delete options that are not relevant.

-   [x] New feature (non-breaking change which adds functionality)
-   [x] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

The change has been tested using a custom COCO dataset which produces the ValueError mentioned in the issue linked above. You may test out the changes by following the guide on the provided Jupyter Notebook using the following resources:
- Google Colab Jupyter Notebook: https://colab.research.google.com/drive/1P57MlFAypF2pvrwIxPQl2hIU8mxoFe7T?usp=sharing
- Custom COCO dataset: https://drive.google.com/file/d/1GHA4z8QK07efOu9Is3v__Dh4V0D483sB/view?usp=sharing

## Any specific deployment considerations

Documentation may require an update.

## Docs

-   [x] Docs updated? What were the changes: Created an entry under `docs/datasets/utils.md` with description and types of the function and its parameters. An example and type have also been added for its output.

DancinParrot committed:
feat: merge multi mask with bitwise or

If a segmentation list of an annotation in a COCO dataset contains more than 1 nested list, seperate binary masks are created for each nested list (polygon). Then, the first polygon is designated as a parent and subsequent polygons in the list will be merged with the parent using numpy logical or.

DancinParrot committed:
test: unit test for merge_polygons (renamed from merge_masks)

The function merge_masks() has been renamed to merge_polygons(). It has been refactored to include explicit function parameters and return value.

DancinParrot committed:
docs: add entry for merge_polygons() function in mkdocs

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/roboflow/supervision?pullRequest=1229) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/roboflow/supervision?pullRequest=1229) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/roboflow/supervision?pullRequest=1229) it.</sub>

SkalskiP commented on supervision/dataset/utils.py:
```diff
@@ -46,6 +47,59 @@ def approximate_mask_with_polygons(
     ]
 
 
+def merge_polygons(
```
I think the `merge_polygons` name is a bit misleading and not conststant with outhe util names in supervision. I think we should call it `polygons_to_mask`.

And of course we already have a function called `polygon_to_mask` that does the same but only take one mask. 

It seems a bit silly that we would have 2 functions doeng almost the same. I thin kwe should:
- rename `merge_polygons` to `polygons_to_mask`.
- make sure that `polygons_to_mask` is not using `polygon_to_mask` internally.
- modify places where `polygon_to_mask` was used and use `polygons_to_mask` instead.
- mask `polygon_to_mask` as deprecated and recommend using `polygons_to_mask` instead.

SkalskiP replied:
@DancinParrot, here is an example of how we mark functions as deprecated. We use `deprecated` decorator to do it.

```
@deprecated(
    "`Detections.from_roboflow` is deprecated and will be removed in "
    "`supervision-0.22.0`. Use `Detections.from_inference` instead."
)
```

SkalskiP replied:
Please use `supervision-0.27.0` as a deprecation version.

SkalskiP commented:
Hi @DancinParrot 👋🏻 Thanks a lot for your contribution 🙏🏻 This is a valid change. As I try to keep the code clean, I left a comment proposing a rename to `polygons_to_mask` and deprecation of the `polygon_to_mask` function. What do you think?

DancinParrot commented:
Hi @SkalskiP ! Thanks for reviewing and providing feedback on my PR!

> As I try to keep the code clean, I left a comment proposing a rename to `polygons_to_mask` and deprecation of the `polygon_to_mask` function. What do you think?

Yup, I think that's a great idea. It'd certainly be a more accurate representation of the function's intended purpose. Since multiple lists of vertices within the "segmentation" field (which in itself is a list) of a COCO dataset annotation is perfectly acceptable, the function should be capable of handling both one and multiple lists without relying on yet another function that introduces another unnecessary layer of abstraction.

I'll proceed with implementing the suggested changes and deprecating the current `polygon_to_mask()` function in favour of `polygons_to_mask()`.

Thanks again for your feedback, appreciate it!

DancinParrot commented:
Hi @SkalskiP ! During refactoring, I noticed a very similar function `_polygons_to_mask()` (see [here](https://github.com/roboflow/supervision/blob/2723735b018cf6f8907d4c706b5f759ff641a53d/supervision/dataset/formats/yolo.py#L43)). The function follows a similar methadology in calling `polygon_to_mask()` for each polygon in the input list, though it does not require the merging of masks into one. In this case, given the similarity of the names with a slight difference in terms of functionality, would the introduction of `polygons_to_mask()` cause confusion?

I suppose we could refactor it in such a way but it does seem to be a bit odd:
```Python
def _polygons_to_masks(
    polygons: List[np.ndarray], resolution_wh: Tuple[int, int]
) -> np.ndarray:
    return np.array(
        [
            polygons_to_mask(polygon=[polygon], resolution_wh=resolution_wh)
            for polygon in polygons
        ],
        dtype=bool,
    )
```

DancinParrot commented:
Hi @SkalskiP ! Any updates on the above?

DancinParrot commented:
Hi @SkalskiP, I noticed the issue still remains unfixed. Is this PR still valid?

Abdul-Mukit commented:
I also need this solution. Can't load the valid datasets without it.

Abdul-Mukit commented:
Hi, @SkalskiP. We would appreciate it if you could comment whether `polygons_to_mask` vs `_polygons_to_mask` creates any confusion, as asked by @DancinParrot.
https://github.com/roboflow/supervision/pull/1229#issuecomment-2133661718

Hi, @DancinParrot. If SkalskiP can't reply, would you have time to implement his original suggestion of `polygons_to_mask` anyway? I personally think that the name is intuitive and clear. Once reviews are addressed, we can request him again.

DancinParrot commented:
> Hi, @DancinParrot. If SkalskiP can't reply, would you have time to implement his original suggestion of `polygons_to_mask` anyway? I personally think that the name is intuitive and clear. Once reviews are addressed, we can request him again.

Sure thing, I'll implement the requested changes while waiting for his reply. Since this is a rather old issue, do give me some time to work on it. Not sure if they are any conflicts with all the commits made thus far.

DancinParrot committed:
fix: merge with upstream and fix imports since polygon_to_mask has been moved from utils.py to converters.py

DancinParrot committed:
refactor: address requested changes to rename merge_polygons

This commit addresses part of the requested changes:
-rename merge polygons to polygons_to_mask
-polygons_to_mask is not using polygon_to_mask internally
-deprecate polygon_to_mask
Further changes required to replace all instances of polygon_to_mask with polygons_to_mask

DancinParrot committed:
refactor: replace all polygon_to_mask with polygons_to_mask

DancinParrot committed:
fix: incorrect list format in coco dataset loader

DancinParrot commented:
Hi @SkalskiP, I have addressed the requested changes and updated my branch with the latest commits, please review it again and let me know if there's any issues. I'm not sure what version to use for deprecation so I decided to stick with the initial one given by you as a placeholder, `supervision-0.27.0`. Changes were tested using the Google Colab notebook and unit tests are all green.

@Abdul-Mukit for your reference as well.

DancinParrot committed:
Merge branch 'develop' into feat/merge_multi_polygons_coco

Borda commented:
@DancinParrot thank you for your effort, but this was already shipped in #1086 🦝
